upstairs/src/volume.rs

```diff
@@ -1092,15 +1094,355 @@ impl Volume {
+                    ReplaceRequestInvalid,
+                    "Cannot replace Region VCR"
+                )
+            }
+
+            VolumeConstructionRequest::File { .. } => {
+                crucible_bail!(ReplaceRequestInvalid, "Cannot replace File VCR")
+            }
+        };
+
+        let n_volume = match replacement {
+            VolumeConstructionRequest::Volume(vcr_v) => vcr_v,
+            VolumeConstructionRequest::Url { .. } => {
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "Replacement URL VCR invalid "
+                )
+            }
+
+            VolumeConstructionRequest::Region { .. } => {
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "Replacement Region VCR invalid"
+                )
+            }
+            VolumeConstructionRequest::File { .. } => {
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "Replacement File VCR invalid"
+                )
+            }
+        };
+
+        if o_volume.id != n_volume.id {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "ID mismatch {} vs. {}",
+                o_volume.id,
+                n_volume.id
+            );
+        }
+        if o_volume.block_size != n_volume.block_size {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "block_size mismatch {} vs. {}",
+                o_volume.block_size,
+                n_volume.block_size
+            )
+        }
+
+        // For a read only parent:
+        // If we had one originally, then we can either have the same
+        // one in the new VCR, or None.  We can't go from None to Some.
+        if n_volume.read_only_parent.is_some() {
+            if o_volume.read_only_parent.is_none() {
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "Unexpected read_only_parent on new VCR"
+                )
+            }
+            if o_volume.read_only_parent != n_volume.read_only_parent {
+                info!(log, "rop mismatch");
+                info!(log, "rop old {:?}", o_volume.read_only_parent);
+                info!(log, "rop new {:?}", n_volume.read_only_parent);
+                crucible_bail!(
+                    ReplaceRequestInvalid,
+                    "read_only_parent mismatch"
+                )
+            }
+        }
+
+        // Sub volume lengths should be the same.
+        if n_volume.sub_volumes.len() != o_volume.sub_volumes.len() {
+            crucible_bail!(
+                ReplaceRequestInvalid,
+                "subvolume len mismatch {} vs. {}",
+                o_volume.sub_volumes.len(),
+                n_volume.sub_volumes.len(),
+            )
+        }
+
+        // Presently, we only support one sub_volume for replacement.
+        if n_volume.sub_volumes.len() != 1 {
```

Comment:
This check doesn't compare `n_volume` to `o_volume` like the others nearby. Is there (or should there be) a separate routine that checks the validity of an individual VCR? (I could also be misunderstanding--is having more than one `sub_volume` OK in general but invalid for replacement? If that's the case I might elaborate on why that is in the comment here.)

Reply:
The original idea was that Volumes could have multiple SubVolumes for different VA ranges, so most of the Volume code just treats the `sub_volume` as a `Vec` and walks all the elements acting on each one.  
However, a few other dependencies have crept in (like activation handling) that now prevent us from supporting more than one `sub_voume` for a `Volume`.  
I wanted to be explicit here and check for a Volume with more than one `sub_volume` and fail it, as this code as written will not handle it.  If we decide in the future to come back and allow it, then we can write the replacement code to handle it then (along with the other places that need updating to support it).

I can leave a comment/bread-crumb here though to better describe that decision to make it easier for future us to find.

Reply:
Makes sense to me--thanks for the context!